File: splitmerge/sorters.py
"""Constructive no-reversal merge sorters for the operation-count model.

Each sorter takes a deck (a permutation written bottom-to-top, the
``machine.py`` convention: index 0 is the bottom, index -1 the top) and returns
the list of move names (``"SA"``/``"SB"``/``"MA"``/``"MB"``) that sort it into
``(1, ..., n)`` with both buffers empty.  Replaying the returned moves on a
fresh machine (``machine.apply_moves``) reproduces the sort -- that replay is
what makes "machine-verified" true (see ``tests/test_sorters.py``).

Three sorters, all *adjacent-merge* (they only ever merge two adjacent sorted
runs), differing only in the binary **merge tree** they build over the runs:

    natural_sort   multi-pass natural merge      cost = 2n*ceil(log2 r)
    topdown_sort   adaptive top-down tree        cost = 2*W(tree)
    hutucker_sort  optimal alphabetic tree       cost = 2*W(tree), W minimal

where ``r`` = number of maximal ascending runs and ``W(T) = sum_i s_i*depth_i``
is the weighted external path length (leaf weights = run sizes).  See
``docs/sources/SORTING-BOUNDS.md`` and ``operation-count-theory.md`` (Part I.3).

The realizations keep runs **ascending bottom-to-top (max on top)** on the deck;
a pour onto a buffer reverses a run to min-on-top, and a count-bounded merge
pours two such runs back ascending.  Merges are by *exact count*, never by
empty-stack test, so a merge never digs into an inert run parked beneath it.
"""
import math


def ascending_runs(deck):
    """Sizes of the maximal ascending (bottom-to-top) runs of ``deck``."""
    deck = list(deck)
    if not deck:
        return []
    sizes = []
    cur = 1
    for i in range(1, len(deck)):
        if deck[i] > deck[i - 1]:
            cur += 1
        else:
            sizes.append(cur)
            cur = 1
    sizes.append(cur)
    return sizes


class _Emitter:
    """A tiny live machine: applies moves to ``(D, A, B)`` while recording them,
    so a sorter can decide each merge step from the real top-of-stack values."""

    def __init__(self, deck):
        self.D = list(deck)
        self.A = []
        self.B = []
        self.moves = []

    def SA(self):
        self.A.append(self.D.pop())
        self.moves.append("SA")

    def SB(self):
        self.B.append(self.D.pop())
        self.moves.append("SB")

    def MA(self):
        self.D.append(self.A.pop())
        self.moves.append("MA")

    def MB(self):
        self.D.append(self.B.pop())
        self.moves.append("MB")

    def _merge_AB(self, n_from_a, n_from_b):
        """Merge the top ``n_from_a`` cards of A with the top ``n_from_b`` of B
        back onto D.  Both runs have their minimum on top, so taking the smaller
        top first emits an ascending run (max on top) of ``n_from_a+n_from_b``
        cards, touching nothing parked deeper."""
        while n_from_a and n_from_b:
            if self.A[-1] < self.B[-1]:
                self.MA()
                n_from_a -= 1
            else:
                self.MB()
                n_from_b -= 1
        while n_from_a:
            self.MA()
            n_from_a -= 1
        while n_from_b:
            self.MB()
            n_from_b -= 1


# ---------------------------------------------------------------------------
# Algorithm 1: multi-pass natural merge.  cost = 2n*ceil(log2 r).
# ---------------------------------------------------------------------------

def natural_sort(deck):
    """Bottom-up two-way natural merge in synchronous passes.  Returns the move
    list; cost is exactly ``2n*ceil(log2 r)`` (``r`` = ascending runs)."""
    m = _Emitter(deck)
    runs = ascending_runs(m.D)
    while len(runs) > 1:
        # distribute: pop runs off the top of D, alternating A, B
        a_runs, b_runs, to_a = [], [], True
        for L in reversed(runs):              # topmost run first
            if to_a:
                for _ in range(L):
                    m.SA()
                a_runs.append(L)
            else:
                for _ in range(L):
                    m.SB()
                b_runs.append(L)
            to_a = not to_a
        # merge: pair the top run of A with the top run of B
        new_runs = []
        while a_runs and b_runs:
            la = a_runs.pop()
            lb = b_runs.pop()
            m._merge_AB(la, lb)
            new_runs.append(la + lb)
        for L in reversed(a_runs):            # odd leftover on A (only one side)
            for _ in range(L):
                m.MA()
            new_runs.append(L)
        for L in reversed(b_runs):
            for _ in range(L):
                m.MB()
            new_runs.append(L)
        runs = new_runs
    return m.moves


# ---------------------------------------------------------------------------
# Tree sorters: build a merge tree over the runs, then realize it (cost 2*W).
# A tree is ('leaf', size) or ('node', left, right, size).
# ---------------------------------------------------------------------------

def _tree_size(t):
    return t[1] if t[0] == "leaf" else t[3]


def weighted_path_length(tree):
    """``W(T) = sum_i size_i * depth_i`` (root at depth 0)."""
    def rec(t, depth):
        if t[0] == "leaf":
            return t[1] * depth
        return rec(t[1], depth + 1) + rec(t[2], depth + 1)
    return rec(tree, 0)


def build_topdown(run_sizes):
    """Adaptive top-down tree: split the run sequence at the boundary nearest
    the card midpoint (long runs stay shallow)."""
    pre = [0]
    for s in run_sizes:
        pre.append(pre[-1] + s)

    def build(i, j):                          # runs [i, j)
        if j - i == 1:
            return ("leaf", run_sizes[i])
        target = (pre[i] + pre[j]) / 2
        best_k, best_d = i + 1, None
        for k in range(i + 1, j):
            d = abs(pre[k] - target)
            if best_d is None or d < best_d:
                best_d, best_k = d, k
        left, right = build(i, best_k), build(best_k, j)
        return ("node", left, right, pre[j] - pre[i])

    if not run_sizes:
        return ("leaf", 0)
    return build(0, len(run_sizes))


def build_hutucker(run_sizes):
    """Optimal alphabetic (order-constrained) merge tree via the optimal-BST
    DP: ``C[i][j]`` = min weight spanning runs ``i..j``.  ``O(r^3)``; this is
    the minimum-``W`` tree, so ``hutucker_sort`` is the optimal no-reversal
    adjacent-merge sorter."""
    r = len(run_sizes)
    if r == 0:
        return ("leaf", 0)
    if r == 1:
        return ("leaf", run_sizes[0])
    pre = [0]
    for s in run_sizes:
        pre.append(pre[-1] + s)
    INF = float("inf")
    C = [[0] * r for _ in range(r)]
    K = [[i for _ in range(r)] for i in range(r)]
    for length in range(2, r + 1):
        for i in range(0, r - length + 1):
            j = i + length - 1
            w = pre[j + 1] - pre[i]
            best, bk = INF, i
            for k in range(i, j):
                v = C[i][k] + C[k + 1][j]
                if v < best:
                    best, bk = v, k
            C[i][j] = best + w
            K[i][j] = bk

    def build(i, j):                          # inclusive runs i..j
        if i == j:
            return ("leaf", run_sizes[i])
        k = K[i][j]
        left, right = build(i, k), build(k + 1, j)
        return ("node", left, right, pre[j + 1] - pre[i])

    return build(0, r - 1)


def realize_tree(deck, tree):
    """Emit the moves that sort ``deck`` by realizing ``tree`` (cost ``2*W``).

    On entry to a node its run-block occupies the top of D, with the right
    (upper-index) child on top.  Sort the right child, park it onto A; sort the
    left child, park it onto B; count-merge the two parked runs back to D.  Each
    node spends ``2 * (cards in node)`` moves, summing to ``2*W``."""
    m = _Emitter(deck)

    def rec(t):
        if t[0] == "leaf":
            return                            # a single ascending run already on D
        _, left, right, _ = t
        rec(right)                            # sort the upper (top-of-D) child
        for _ in range(_tree_size(right)):    # park it onto A (reverses)
            m.SA()
        rec(left)                             # sort the lower child (now on top)
        for _ in range(_tree_size(left)):     # park it onto B
            m.SB()
        m._merge_AB(_tree_size(right), _tree_size(left))

    rec(tree)
    return m.moves


def topdown_sort(deck):
    """Adaptive top-down merge sort; returns the move list (cost ``2*W``)."""
    return realize_tree(deck, build_topdown(ascending_runs(deck)))


def hutucker_sort(deck):
    """Optimal Hu-Tucker merge sort; returns the move list (cost ``2*W``, ``W``
    minimal over all order-preserving merge trees)."""
    return realize_tree(deck, build_hutucker(ascending_runs(deck)))


# ---------------------------------------------------------------------------
# Closed-form costs (no machine replay needed), for cross-checking.
# ---------------------------------------------------------------------------

def natural_cost(deck):
    """``2n*ceil(log2 r)`` -- the exact natural_sort move count."""
    n = len(deck)
    r = len(ascending_runs(deck))
    return 0 if r <= 1 else 2 * n * math.ceil(math.log2(r))


def topdown_cost(deck):
    return 2 * weighted_path_length(build_topdown(ascending_runs(deck)))


def hutucker_cost(deck):
    return 2 * weighted_path_length(build_hutucker(ascending_runs(deck)))

File: src/sorters.rs
//! Constructive no-reversal merge sorters that emit real machine moves (verified
//! by replay). Three adjacent-merge sorters differing only in the merge tree:
//! `natural` (multi-pass, `2n*ceil(log2 r)`), `top_down` and `hu_tucker`
//! (`2*W(tree)`, `W` minimal for Hu-Tucker). See docs/NOTES.md §I.3.

use crate::machine::{Card, Move};

/// Sizes of the maximal ascending (bottom-to-top) runs of `deck`.
pub fn ascending_runs(deck: &[Card]) -> Vec<usize> {
    if deck.is_empty() {
        return Vec::new();
    }
    let mut sizes = Vec::new();
    let mut cur = 1;
    for i in 1..deck.len() {
        if deck[i] > deck[i - 1] {
            cur += 1;
        } else {
            sizes.push(cur);
            cur = 1;
        }
    }
    sizes.push(cur);
    sizes
}

/// A tiny live machine that records the moves it makes.
struct Emitter {
    d: Vec<Card>,
    a: Vec<Card>,
    b: Vec<Card>,
    moves: Vec<Move>,
}

impl Emitter {
    fn new(deck: &[Card]) -> Self {
        Emitter {
            d: deck.to_vec(),
            a: Vec::new(),
            b: Vec::new(),
            moves: Vec::new(),
        }
    }
    fn sa(&mut self) {
        let c = self.d.pop().unwrap();
        self.a.push(c);
        self.moves.push(Move::SA);
    }
    fn sb(&mut self) {
        let c = self.d.pop().unwrap();
        self.b.push(c);
        self.moves.push(Move::SB);
    }
    fn ma(&mut self) {
        let c = self.a.pop().unwrap();
        self.d.push(c);
        self.moves.push(Move::MA);
    }
    fn mb(&mut self) {
        let c = self.b.pop().unwrap();
        self.d.push(c);
        self.moves.push(Move::MB);
    }
    /// Merge the top `na` of A with the top `nb` of B back onto D (both runs have
    /// their minimum on top), by exact count.
    fn merge_ab(&mut self, mut na: usize, mut nb: usize) {
        while na > 0 && nb > 0 {
            if *self.a.last().unwrap() < *self.b.last().unwrap() {
                self.ma();
                na -= 1;
            } else {
                self.mb();
                nb -= 1;
            }
        }
        while na > 0 {
            self.ma();
            na -= 1;
        }
        while nb > 0 {
            self.mb();
            nb -= 1;
        }
    }
}

/// Multi-pass natural merge. Cost exactly `2n*ceil(log2 r)`.
pub fn natural_sort(deck: &[Card]) -> Vec<Move> {
    let mut m = Emitter::new(deck);
    let mut runs = ascending_runs(&m.d);
    while runs.len() > 1 {
        let (mut a_runs, mut b_runs) = (Vec::new(), Vec::new());
        let mut to_a = true;
        for &len in runs.iter().rev() {
            if to_a {
                for _ in 0..len {
                    m.sa();
                }
                a_runs.push(len);
            } else {
                for _ in 0..len {
                    m.sb();
                }
                b_runs.push(len);
            }
            to_a = !to_a;
        }
        let mut new_runs = Vec::new();
        while let (Some(la), Some(lb)) = (a_runs.last().copied(), b_runs.last().copied()) {
            a_runs.pop();
            b_runs.pop();
            m.merge_ab(la, lb);
            new_runs.push(la + lb);
        }
        for &len in a_runs.iter().rev() {
            for _ in 0..len {
                m.ma();
            }
            new_runs.push(len);
        }
        for &len in b_runs.iter().rev() {
            for _ in 0..len {
                m.mb();
            }
            new_runs.push(len);
        }
        runs = new_runs;
    }
    m.moves
}

/// A binary merge tree over the runs.
pub enum Tree {
    Leaf(usize),
    Node(Box<Tree>, Box<Tree>, usize),
}

impl Tree {
    pub fn size(&self) -> usize {
        match self {
            Tree::Leaf(s) => *s,
            Tree::Node(_, _, s) => *s,
        }
    }
}

/// `W(T) = sum_i size_i * depth_i` (root at depth 0).
pub fn weighted_path_length(t: &Tree) -> usize {
    fn rec(t: &Tree, depth: usize) -> usize {
        match t {
            Tree::Leaf(s) => s * depth,
            Tree::Node(l, r, _) => rec(l, depth + 1) + rec(r, depth + 1),
        }
    }
    rec(t, 0)
}

fn prefix_sums(run_sizes: &[usize]) -> Vec<usize> {
    let mut pre = vec![0usize; run_sizes.len() + 1];
    for (i, &s) in run_sizes.iter().enumerate() {
        pre[i + 1] = pre[i] + s;
    }
    pre
}

/// Adaptive top-down tree: split at the boundary nearest the card midpoint.
pub fn build_topdown(run_sizes: &[usize]) -> Tree {
    if run_sizes.is_empty() {
        return Tree::Leaf(0);
    }
    let pre = prefix_sums(run_sizes);
    fn build(i: usize, j: usize, run_sizes: &[usize], pre: &[usize]) -> Tree {
        if j - i == 1 {
            return Tree::Leaf(run_sizes[i]);
        }
        let target = (pre[i] + pre[j]) as f64 / 2.0;
        let mut best_k = i + 1;
        let mut best_d = f64::INFINITY;
        for k in (i + 1)..j {
            let d = (pre[k] as f64 - target).abs();
            if d < best_d {
                best_d = d;
                best_k = k;
            }
        }
        let left = build(i, best_k, run_sizes, pre);
        let right = build(best_k, j, run_sizes, pre);
        Tree::Node(Box::new(left), Box::new(right), pre[j] - pre[i])
    }
    build(0, run_sizes.len(), run_sizes, &pre)
}

/// Optimal alphabetic (order-constrained) merge tree via the optimal-BST DP.
pub fn build_hutucker(run_sizes: &[usize]) -> Tree {
    let r = run_sizes.len();
    if r == 0 {
        return Tree::Leaf(0);
    }
    if r == 1 {
        return Tree::Leaf(run_sizes[0]);
    }
    let pre = prefix_sums(run_sizes);
    let mut c = vec![vec![0usize; r]; r];
    let mut kbest = vec![vec![0usize; r]; r];
    for i in 0..r {
        kbest[i][i] = i;
    }
    for length in 2..=r {
        for i in 0..=(r - length) {
            let j = i + length - 1;
            let w = pre[j + 1] - pre[i];
            let mut best = usize::MAX;
            let mut bk = i;
            for k in i..j {
                let v = c[i][k] + c[k + 1][j];
                if v < best {
                    best = v;
                    bk = k;
                }
            }
            c[i][j] = best + w;
            kbest[i][j] = bk;
        }
    }
    fn build(i: usize, j: usize, run_sizes: &[usize], pre: &[usize], kbest: &[Vec<usize>]) -> Tree {
        if i == j {
            return Tree::Leaf(run_sizes[i]);
        }
        let k = kbest[i][j];
        let left = build(i, k, run_sizes, pre, kbest);
        let right = build(k + 1, j, run_sizes, pre, kbest);
        Tree::Node(Box::new(left), Box::new(right), pre[j + 1] - pre[i])
    }
    build(0, r - 1, run_sizes, &pre, &kbest)
}

/// Emit the moves that sort `deck` by realizing `tree` (cost `2*W`).
pub fn realize_tree(deck: &[Card], tree: &Tree) -> Vec<Move> {
    let mut m = Emitter::new(deck);
    fn rec(t: &Tree, m: &mut Emitter) {
        if let Tree::Node(l, r, _) = t {
            rec(r, m); // sort the upper (top-of-D) child
            for _ in 0..r.size() {
                m.sa(); // park it onto A
            }
            rec(l, m);
            for _ in 0..l.size() {
                m.sb(); // park onto B
            }
            m.merge_ab(r.size(), l.size());
        }
    }
    rec(tree, &mut m);
    m.moves
}

pub fn topdown_sort(deck: &[Card]) -> Vec<Move> {
    realize_tree(deck, &build_topdown(&ascending_runs(deck)))
}

pub fn hutucker_sort(deck: &[Card]) -> Vec<Move> {
    realize_tree(deck, &build_hutucker(&ascending_runs(deck)))
}

// ----- closed-form costs (no replay) ------------------------------------------

pub fn natural_cost(deck: &[Card]) -> usize {
    let n = deck.len();
    let r = ascending_runs(deck).len();
    if r <= 1 {
        0
    } else {
        2 * n * (usize::BITS - (r - 1).leading_zeros()) as usize // 2n * ceil(log2 r)
    }
}

pub fn topdown_cost(deck: &[Card]) -> usize {
    2 * weighted_path_length(&build_topdown(&ascending_runs(deck)))
}

pub fn hutucker_cost(deck: &[Card]) -> usize {
    2 * weighted_path_length(&build_hutucker(&ascending_runs(deck)))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::machine::State;

    fn perms(n: usize, f: &mut dyn FnMut(&[Card])) {
        let mut p: Vec<Card> = (1..=n as Card).collect();
        permute(&mut p, 0, f);
    }
    fn permute(p: &mut Vec<Card>, i: usize, f: &mut dyn FnMut(&[Card])) {
        if i == p.len() {
            f(p);
            return;
        }
        for j in i..p.len() {
            p.swap(i, j);
            permute(p, i + 1, f);
            p.swap(i, j);
        }
    }

    #[test]
    fn all_perms_sort_and_count() {
        for n in 1..=7 {
            let goal = State::goal(n);
            perms(n, &mut |p| {
                for (sort, cost) in [
                    (natural_sort as fn(&[Card]) -> Vec<Move>, natural_cost as fn(&[Card]) -> usize),
                    (topdown_sort, topdown_cost),
                    (hutucker_sort, hutucker_cost),
                ] {
                    let moves = sort(p);
                    assert_eq!(State::from_deck(p.to_vec()).applied(&moves), goal);
                    assert_eq!(moves.len(), cost(p));
                }
            });
        }
    }

    #[test]
    fn worst_case_reversed_52() {
        let rev: Vec<Card> = (1..=52).rev().collect();
        assert_eq!(natural_cost(&rev), 624);
        assert_eq!(topdown_cost(&rev), 600);
        assert_eq!(hutucker_cost(&rev), 600);
        assert_eq!(
            State::from_deck(rev.clone()).applied(&hutucker_sort(&rev)),
            State::goal(52)
        );
    }
}
